### ISSUE TITLE
Update default Tooltips in Frame5 to support HTML and show instantly.

### ISSUE THREAD

**Issue by FullBleed** [RPTools/maptool#2850]
**Is your feature request related to a problem? Please describe.**
In the old frame, the default tooltip was really quite nice.  It displayed instantly and supported HTML.  The default tooltip in frame5 takes a second or two to pop up (too long) and does not support even basic HTML (plain text only with no word wrap.)  That makes the default frame5 tooltip a significant downgrade (to the point of being kind of useless.)

![frame5 tooltips](https://user-images.githubusercontent.com/37131742/124869503-4f60e680-df8f-11eb-949e-4252b531f6b5.jpg)

**Describe the solution you'd like**
I would very much like it to mimic the old frame styling and features (html, wordwrap, and frame break) for both compatibility and usability.  

**Describe alternatives you've considered**
Frame5 allows you to make your own CSS tooltips but they have significant limitations.  First, I haven't found one that will break the element it's in (like a cell or a table or the frame5 boundry if outside a table.)  This is a HUGE limitation compared to what the old frame tooltip AND (for that matter) what the current default frame tooltip can do.  Second, forcing users to do this isn't very nice and contributes to gating another MT feature to requiring "coding knowledge."  It's not enough to learn how to create the tooltip, you've now got to learn CSS (and even that isn't enough... see below.)

**Additional context**
Here is another screenshot showsing a basic HTML table in the old frame (left) versus in the new frame (right.)  Not the best table but I wanted it to be basic.  In both examples, the table pops up when you hover on the word "Free".  If I build that same table using a CSS tooltip it won't work because the word "Free" is in a table cell and the tooltip is limited by the width of that cell/element and won't expand across the frame (and definitely not the frame.)  So, as is, now we can't even use a decent tooltip in tables.

![frame5 garbage tooltip](https://user-images.githubusercontent.com/37131742/124870360-94394d00-df90-11eb-8bbe-20fd45da18e9.jpg)

**Comment by Azhrei** [comment]
I don't think we can change this.

I looked through the code for `WebView` and associated classes.

`WebView` delegates most rendering to `javafx.web/com.sun.javafx.webkit.WebPageClientImpl`, and that class is where the tooltip content is set.  There is no accommodation made for anything other than text.  (This means the only way to get HTML into a tooltip is via the CSS `content` route.)

Additionally, the event handler that causes the tooltip to appear, `javafx.controls/javafx.scene.control.Tooltip::MOVE_HANDLER()`, determines the window that contains the HTML element that has the tooltip attached, and makes that window the parent for the tooltip itself.  Since all children are automatically clipped to their parent window's dimensions, there is no way for the tooltip to extend beyond the border of the MapTool frame.  (The way around this for dialogs and such is to use the screen as the parent, not the application's top-level window.)

Theoretically, it would be possible to subclass the `Tooltip` class and override the event handler.  However, it still wouldn't work, as the creation of the tooltip object is embedded inside other code that does not allow for the class name to be a parameter.  (Ie., there would be no way to use this new class, even if it were created.)

(For all of the above reasons, I expect tooltips in overlays cannot extend outside the boundaries of the map either.)

In summary, I don't think there's anything MapTool can do about this.  I'll leave it to someone else to close it, since maybe someone will have an idea...

**Comment by FullBleed** [comment]
Out of curiosity, why can the default tooltip break frame?  I assume there is some kind of default styling being used for it.

**Comment by Azhrei** [comment]
The Swing library (used by the rest of MT) sets the parent of the tooltip to be the root window in the containment hierarchy of the component  (or the screen itself in some cases).

**Comment by FullBleed** [comment]
> The Swing library (used by the rest of MT) sets the parent of the tooltip to be the root window in the containment hierarchy of the component (or the screen itself in some cases).

So, to be clear, Swing does something different with frame5() than frame()?

In frame() we have html, light background, and frame breaking.
In frame5() we have text only, black semi-transparent background, white text, and frame breaking.

Both of those behaviours are the result of Swing?

**Comment by Azhrei** [comment]
> So, to be clear, Swing does something different with frame5() than frame()?
> 
Swing handles `frame()`.  JavaFX handles `frame5()`.

Swing has its own HTML parser and CSS implementation (HTML 3.2 and CSS 1+ — remember those?).  JavaFX delegates to WebView which was originally based on Apple's WebKit, and which is now the core rendering engine of multiple desktop browsers.  WebKit continues to evolve while Swing does not.

Since MapTool is primarily Swing, it forwards calls to `frame5()` to an instance of WebView.  Swing gives the WebView it's bounding box (the window panel of the docking pane) and then WebView is responsible for everything inside.

Is it possible for it to work some other way?  Sure, anything is possible.  Swing could create a top-level window and then embed the WebView inside it, but the WebView will still be limited to that window (and you'd still lose the ability to dock panels, which is a feature of the Swing-based JIDE libraries).  No telling what effect it would have on full screen mode, but it could probably be made to work.

Or, Swing could tell JavaFX to create its own top-level window and put the WebView in it.  Now the tooltips could (not sure if they _would_) extend beyond the edge of the window.  But again, you lose the docking panels.  @Jamz looked into multiple different libraries for docking panels and couldn't find any that were (1) robust enough for MT to use, and (2) currently being maintained.  (There are probably other GH issues that reference DockingFX and similar libraries, or maybe it was discussed in Discord...)
—
Frank Edwards
Edwards & Edwards Consulting, LLC
Hours:  10A-6P ET
Phone:  (813) 406-0604 (office, mobile, and SMS texts)

No electrons were harmed in the creation or transmission of this message.

**Comment by aliasmask** [comment]
I just assumed the normal toolip using title="tooltip" didn't work in html5. It's nice to see at least plain text works, although that delay of 1 second is huge. It would be nice have it work like it does for normal frame() as a feature because it's a bit of a pain to set up floating divs and whatnot manually.

**Comment by Azhrei** [comment]
If you want to just change the activation and hide delays, try CSS:  https://docs.oracle.com/javase/9/docs/api/javafx/scene/doc-files/cssref.html#tooltip (I haven’t tested this myself.)

**Comment by FullBleed** [comment]
> If you want to just change the activation and hide delays, try CSS: https://docs.oracle.com/javase/9/docs/api/javafx/scene/doc-files/cssref.html#tooltip (I haven’t tested this myself.)

Interesting.

I couldn't get it to work... maybe someone else can.  @melek ?

**Comment by Azhrei** [comment]
> I couldn't get it to work... maybe someone else can. @melek ?

I've decided those CSS styles won't apply to content in a WebView (the code in JavaFX for displaying HTML), they only apply to JavaFX code written directly by the programmer.  I tried variations, like replacing `-fx-` with `-webkit-` and similar (since WebView is based on the WebKit engine), but no luck.

**Comment by FullBleed** [comment]
@Azhrei @melek Thanks for the follow-up.  I've actually got several different tooltips setup, but not having the frame breaking, and having to use script and css to get a decent tooltip is a pain I was hoping we could avoid.  The original default frame() tooltip was just far nicer and easier to use.

Should we leave this open just in case some WebView guru comes along with a solution?

**Comment by cwisniew** [comment]
Apart from WebView another reason that the tool tip can not break out of the frame5/dialog5 is because WebView is a JavaFX control and MapTool is a Swing program with embedded JavaFX. This is done by creating a Swing control that can embed a JavaFX control, and JavaFX is limited to rendering inside of this (which happens to be inside the frame5/dialog5). So even if it were possible for WebView to render outside of the embedded context it sits in, meaning there is no way we could have the tooltip break out of the frame5/dialog5.

I am going to close this as wontfix because we cant fix it (well if we went down track of browser based client it wouldn't be a problem but that is a lost cause see #3419
